patch_api/rest.py

```diff
@@ -110,6 +111,58 @@ def __init__(self, api_key, configuration, pools_size=4, maxsize=None):
                 **addition_pool_args
             )
 
+    def recursive_urlencode(self, d):
+        """URL-encode a multidimensional dictionary.
+
+        >>> data = {'a': 'b&c', 'd': {'e': {'f&g': 'h*i'}}, 'j': 'k'}
+        >>> recursive_urlencode(data)
+        u'a=b%26c&j=k&d[e][f%26g]=h%2Ai'
+        """
+
+        def recursion(d, base=[]):
+            pairs = []
+
+            for key, value in d.items():
+                new_base = base + [key]
+                if hasattr(value, "values"):
+                    pairs += recursion(value, new_base)
+                else:
+                    new_pair = None
+                    if len(new_base) > 1:
+                        first = urllib.parse.quote(new_base.pop(0))
+                        rest = map(lambda x: urllib.parse.quote(x), new_base)
+                        new_pair = "%s[%s]=%s" % (
+                            first,
+                            "][".join(rest),
+                            urllib.parse.quote(str(value)),
+                        )
+                    else:
+                        new_pair = "%s=%s" % (
+                            urllib.parse.quote(str(key)),
+                            urllib.parse.quote(str(value)),
+                        )
+                    pairs.append(new_pair)
+            return pairs
+
+        return "&".join(recursion(d))
+
+    def encoded_query_params(self, query_params):
+        if not query_params:
+            return ""
+
+        final_query_params = ""
+        for key, value in query_params:
+            if isinstance(value, dict):
+                nested_param = {}
+                nested_param[key] = value
+                final_query_params += self.recursive_urlencode(nested_param)
+                query_params.remove((key, value))
+
+        if query_params:
+            final_query_params += "&" + urlencode(query_params)
+
+        return "?" + final_query_params
```

Comment:
this is all custom code. I took `recursive_urlencode` from a stackoverflow answer and modified it to work with python3. 
`encoded_query_params` is a method I wrote

Reply:
Is there a chance we could unit test those two methods? I trust the global test but this might save some pain later.